plugins/flytekit-inference/flytekitplugins/inference/ollama/serve.py

```diff
@@ -0,0 +1,187 @@
+import base64
+from dataclasses import dataclass
+from typing import Optional
+
+from flytekit.configuration.default_images import DefaultImages
+
+from ..sidecar_template import ModelInferenceTemplate
+
+
+@dataclass
+class Model:
+    """Represents the configuration for a model used in a Kubernetes pod template.
+
+    :param name: The name of the model.
+    :param mem: The amount of memory allocated for the model, specified as a string. Default is "500Mi".
+    :param cpu: The number of CPU cores allocated for the model. Default is 1.
+    :param modelfile: The actual model file as a JSON-serializable string. This represents the file content. Default is `None` if not applicable.
+    """
+
+    name: str
+    mem: str = "500Mi"
+    cpu: int = 1
+    modelfile: Optional[str] = None
+
+
+class Ollama(ModelInferenceTemplate):
+    def __init__(
+        self,
+        *,
+        model: Model,
+        image: str = "ollama/ollama",
+        port: int = 11434,
+        cpu: int = 1,
+        gpu: int = 1,
+        mem: str = "15Gi",
+    ):
+        """Initialize Ollama class for managing a Kubernetes pod template.
+
+        :param model: An instance of the Model class containing the model's configuration, including its name, memory, CPU, and file.
+        :param image: The Docker image to be used for the container. Default is "ollama/ollama".
+        :param port: The port number on which the container should expose its service. Default is 11434.
+        :param cpu: The number of CPU cores requested for the container. Default is 1.
+        :param gpu: The number of GPUs requested for the container. Default is 1.
+        :param mem: The amount of memory requested for the container, specified as a string. Default is "15Gi".
+        """
+        self._model_name = model.name
+        self._model_mem = model.mem
+        self._model_cpu = model.cpu
+        self._model_modelfile = model.modelfile
+
+        super().__init__(image=image, port=port, cpu=cpu, gpu=gpu, mem=mem)
+
+        self.setup_ollama_pod_template()
+
+    def setup_ollama_pod_template(self):
+        from kubernetes.client.models import (
+            V1Container,
+            V1ResourceRequirements,
+            V1SecurityContext,
+        )
+
+        container_name = "create-model" if self._model_modelfile else "pull-model"
+
+        if self._model_modelfile:
+            base_code = """
+import base64
+import ollama
+"""
+            encoded_modelfile = base64.b64encode(self._model_modelfile.encode("utf-8")).decode("utf-8")
+
+            if "{inputs" in self._model_modelfile:
+                python_code = f"""
+{base_code}
+import os
```

Comment:
A lot of the complexity of this PR comes from having boilerplate code to download inputs. Do you think it's worth having a separate image that just download inputs in PodTemplate + init containers?

Reply:
i chose this approach to avoid maintaining a separate image. i tested it, and it works fine. do you think it would be better to move the code into an image for better readability? @thomasjpfan

Reply:
how would that work?

Right now we have one sidecar container (which runs the ollama server in this PR) and one init container that waits 15 seconds before it runs the python code.

Downloading the files doesn't need to wait for the ollama server to start, right? And since init containers [run serially](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#:~:text=If%20you%20specify%20multiple%20init,and%20runs%20them%20as%20usual.), we could in principle setup an init container that just downloads the necessary files and a subsequent init container that creates / pulls the model (which depends on the files to exist).

Reply:
that’s one approach we could take, sure. we’d need to configure a volume and mount it in both init containers for it to work. do you think i should create a separate image for the inputs container? since it's safe to include long arguments, is it fine to just include the code in args? @eapolinario

Reply:
For me, it comes down to "Is is generally useful to download input data in context of a PodTemplate?". If we are going to build more init containers that needs to download input data, then I think it's worth abstracting away the "Downloading input data" part into it's own init container. For example, downloading input data would also be useful for:

- NIM
- vLLM
- TorchServe
- TensorRT
- BentoML

As for this PR, I do not think this point is blocking.